src/simulators/multi_state_executor.hpp

```diff
@@ -0,0 +1,758 @@
+ * that they have been altered from the originals.
+ */
+
+#ifndef _multi_state_executor_hpp_
+#define _multi_state_executor_hpp_
+
+#include "simulators/circuit_executor.hpp"
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#ifdef AER_MPI
+#include <mpi.h>
+#endif
+
+#include "simulators/shot_branching.hpp"
+
+namespace AER {
+
+namespace CircuitExecutor {
+
+//-------------------------------------------------------------------------
+// Multiple-shots executor class implementation
+//-------------------------------------------------------------------------
+template <class state_t>
+class MultiStateExecutor : public Executor<state_t> {
+  using Base = Executor<state_t>;
+
+protected:
+  std::vector<state_t> states_;
+  std::vector<ClassicalRegister> cregs_; // classical registers for all shots
+
+  // number of qubits for the circuit
+  uint_t num_qubits_;
+
+  uint_t num_global_states_; // number of total shots
+  uint_t num_local_states_;  // number of local shots
+
+  uint_t global_state_index_; // beginning chunk index for this process
+  reg_t state_index_begin_;   // beginning chunk index for each process
+  reg_t state_index_end_;     // ending chunk index for each process
+  uint_t num_active_states_;  // number of active shots in current loop
+
+  bool shot_omp_parallel_; // using thread parallel to process loop of chunks or
+                           // not
+
+  bool set_parallelization_called_ =
+      false; // this flag is used to check set_parallelization is already
+             // called, if yes the call sets max_batched_shots_
+  uint_t num_max_shots_ =
+      1; // max number of shots can be stored on available memory
+
+  int max_matrix_qubits_; // max qubits for matrix
+
+  // shot branching
+  bool shot_branching_enable_ = true;
+  bool runtime_noise_sampling_enable_ = false;
+
+  // group of states (GPU devices)
+  uint_t num_groups_; // number of groups of chunks
+  reg_t top_state_of_group_;
+  reg_t num_states_in_group_;
+  int num_threads_per_group_; // number of outer threads per group
+
+  uint_t num_creg_memory_ =
+      0; // number of total bits for creg (reserve for multi-shots)
+  uint_t num_creg_registers_ = 0;
+
+  // OpenMP qubit threshold
+  int omp_qubit_threshold_ = 14;
+
+  // Threshold for chopping small values to zero in JSON
+  double json_chop_threshold_ = 1e-10;
+
+  // Set a global phase exp(1j * theta) for the state
+  bool has_global_phase_ = false;
+  complex_t global_phase_ = 1;
+
+public:
+  MultiStateExecutor();
+  virtual ~MultiStateExecutor();
+
+  size_t required_memory_mb(const Circuit &circuit,
+                            const Noise::NoiseModel &noise) const override {
+    state_t tmp;
+    return tmp.required_memory_mb(circuit.num_qubits, circuit.ops);
+  }
+
+  uint_t get_process_by_chunk(uint_t cid);
+
+protected:
+  void set_config(const Config &config) override;
+
+  void set_distribution(uint_t num_states);
```

Comment:
Could you add comments to explain this method? Originally, I found the following comment in`state_chunk.hpp` to the method of the same name excepting its argument name `uint_t nprocs`.

> // set number of processes to be distributed